File: Assets/Scripts/weapons/Gun.js
#pragma strict

/**
 * @author: cmdr2 <secondary.cmdr2@gmail.com>
 */

private static var instance : Gun;

/* dependencies */
public var Bullet : GameObject;
public var muzzleFlash : ParticleSystem;
public var light1 : GameObject;
public var light2 : GameObject;
public var light3 : GameObject;


/* globals */
private var FIRE_RATE : float = 0.232 * 3;
private var MIN_FIRE_SOUND_DURATION : float = 0.232 * 3; // seconds
private var SIGHTLINE_DISTANCE : float = 30; // meters


/* scratchpad */
private var gameOver : boolean = false;

private var nextFire : float = 0.0;
private var fireSoundDuration : float = MIN_FIRE_SOUND_DURATION;

private var fireSoundStartTime : float = 0.0;

private var muzzleTimer : float = 0.0;
private var muzzleCooler : float = 0.1;
private var bulletQueueCount : int = 0;
private var isFiring : boolean = false;


function Awake() {
	MuzzleFlash(false);
}


public static function Instance() : Gun {
	if (instance == null) {
		instance = UnityEngine.Object.FindObjectOfType(Gun);
	}
	return instance;
}


@script RequireComponent(AudioSource)
function Update () {
	if (gameOver) {
		if (GetComponent.<AudioSource>()) {
			if (GetComponent.<AudioSource>().isPlaying) GetComponent.<AudioSource>().Stop();
		}
	} else {
		var triggered : boolean = Cardboard.SDK.CardboardTriggered || Input.GetButton("Fire1");
		if (GameManager.Instance().gameState == GameState.IN_LEVEL && triggered && Time.time > nextFire && Time.time > fireSoundDuration) {
			nextFire = Time.time + FIRE_RATE * 0.3;
			Fire();
		}
		
		if (Input.GetButtonUp("Fire1")) {
			bulletQueueCount = 3;
		}
	}
}


private function MuzzleFlash(on : boolean) {
	if (on) {
		if (muzzleFlash) muzzleFlash.Emit(100);
	} else {
		if (muzzleFlash) muzzleFlash.Stop();// = false;
	}
	if (light1) light1.SetActive(on);
	if (light2) light2.SetActive(on);
	if (light3) light3.SetActive(on);
}


private function Fire() {
	if (isFiring) {
		bulletQueueCount += 3;
	} else {
		isFiring = true;
		if (GetComponent.<AudioSource>()) GetComponent.<AudioSource>().Play();
		bulletQueueCount = 9;
		while (bulletQueueCount > 0) {
			if (gameOver) {
				bulletQueueCount = 0;
				break;
			}
			bulletQueueCount--;
			MuzzleFlash(true);
			Instantiate(Bullet, transform.position, transform.rotation);
			yield WaitForSeconds(0.03);
			MuzzleFlash(false);
			
			yield WaitForSeconds(0.04);
		}
		isFiring = false;
		if (GetComponent.<AudioSource>()) GetComponent.<AudioSource>().Stop();
	}
}


private function StartGunSound() {
	//  if still playing previous single-fire "3 bullet" sound
	if (Time.time - fireSoundStartTime < fireSoundDuration) {
		fireSoundStartTime = Time.time;
	} else {
		fireSoundStartTime = Time.time;
		fireSoundDuration = MIN_FIRE_SOUND_DURATION;
		
		if (GetComponent.<AudioSource>()) GetComponent.<AudioSource>().Play();
	}
}


private function StopGunSound() {
	// play atleast 3 bullet sounds for single-fire
	var timeLeft = fireSoundDuration - (Time.time - fireSoundStartTime);
	while ( timeLeft > 0 ) {
		yield WaitForSeconds(timeLeft);
		timeLeft = fireSoundDuration - (Time.time - fireSoundStartTime);
	}
	
	
	if (GetComponent.<AudioSource>()) GetComponent.<AudioSource>().Stop();
}


public function GameOver() {
	GameOver(true);
}


public function GameOver(over) {
	gameOver = over;
}

File: Assets/Scripts/weapons/Gun.cs
using UnityEngine;
using System.Collections;

[System.Serializable]
/**
 * @author: cmdr2 <secondary.cmdr2@gmail.com>
 */
/* dependencies */
/* globals */
 // seconds
 // meters
/* scratchpad */
[UnityEngine.RequireComponent(typeof(AudioSource))]
public partial class Gun : MonoBehaviour
{
    private static Gun instance;
    public GameObject Bullet;
    public ParticleSystem muzzleFlash;
    public GameObject light1;
    public GameObject light2;
    public GameObject light3;
    private float FIRE_RATE;
    private float MIN_FIRE_SOUND_DURATION;
    private float SIGHTLINE_DISTANCE;
    private bool gameOver;
    private float nextFire;
    private float fireSoundDuration;
    private float fireSoundStartTime;
    private float muzzleTimer;
    private float muzzleCooler;
    private int bulletQueueCount;
    private bool isFiring;
    public virtual void Awake()
    {
        this.MuzzleFlash(false);
    }

    public static Gun Instance()
    {
        if (Gun.instance == null)
        {
            Gun.instance = (Gun) UnityEngine.Object.FindObjectOfType(typeof(Gun));
        }
        return Gun.instance;
    }

    public virtual void Update()
    {
        if (this.gameOver)
        {
            if (this.GetComponent<AudioSource>())
            {
                if (this.GetComponent<AudioSource>().isPlaying)
                {
                    this.GetComponent<AudioSource>().Stop();
                }
            }
        }
        else
        {
            bool triggered = Cardboard.SDK.CardboardTriggered || Input.GetButton("Fire1");
            if ((((GameManager.Instance().gameState == GameState.IN_LEVEL) && triggered) && (Time.time > this.nextFire)) && (Time.time > this.fireSoundDuration))
            {
                this.nextFire = Time.time + (this.FIRE_RATE * 0.3f);
                this.StartCoroutine(this.Fire());
            }
            if (Input.GetButtonUp("Fire1"))
            {
                this.bulletQueueCount = 3;
            }
        }
    }

    private void MuzzleFlash(bool on)
    {
        if (on)
        {
            if (this.muzzleFlash)
            {
                this.muzzleFlash.Emit(100);
            }
        }
        else
        {
            if (this.muzzleFlash)
            {
                this.muzzleFlash.Stop();// = false;
            }
        }
        if (this.light1)
        {
            this.light1.SetActive(on);
        }
        if (this.light2)
        {
            this.light2.SetActive(on);
        }
        if (this.light3)
        {
            this.light3.SetActive(on);
        }
    }

    private IEnumerator Fire()
    {
        if (this.isFiring)
        {
            this.bulletQueueCount = this.bulletQueueCount + 3;
        }
        else
        {
            this.isFiring = true;
            if (this.GetComponent<AudioSource>())
            {
                this.GetComponent<AudioSource>().Play();
            }
            this.bulletQueueCount = 9;
            while (this.bulletQueueCount > 0)
            {
                if (this.gameOver)
                {
                    this.bulletQueueCount = 0;
                    break;
                }
                this.bulletQueueCount--;
                this.MuzzleFlash(true);
                UnityEngine.Object.Instantiate(this.Bullet, this.transform.position, this.transform.rotation);
                yield return new WaitForSeconds(0.03f);
                this.MuzzleFlash(false);
                yield return new WaitForSeconds(0.04f);
            }
            this.isFiring = false;
            if (this.GetComponent<AudioSource>())
            {
                this.GetComponent<AudioSource>().Stop();
            }
        }
    }

    private void StartGunSound()
    {
        //  if still playing previous single-fire "3 bullet" sound
        if ((Time.time - this.fireSoundStartTime) < this.fireSoundDuration)
        {
            this.fireSoundStartTime = Time.time;
        }
        else
        {
            this.fireSoundStartTime = Time.time;
            this.fireSoundDuration = this.MIN_FIRE_SOUND_DURATION;
            if (this.GetComponent<AudioSource>())
            {
                this.GetComponent<AudioSource>().Play();
            }
        }
    }

    private IEnumerator StopGunSound()
    {
        // play atleast 3 bullet sounds for single-fire
        float timeLeft = this.fireSoundDuration - (Time.time - this.fireSoundStartTime);
        while (timeLeft > 0)
        {
            yield return new WaitForSeconds(timeLeft);
            timeLeft = this.fireSoundDuration - (Time.time - this.fireSoundStartTime);
        }
        if (this.GetComponent<AudioSource>())
        {
            this.GetComponent<AudioSource>().Stop();
        }
    }

    public virtual void GameOver()
    {
        this.GameOver(true);
    }

    public virtual void GameOver(object over)
    {
        this.gameOver = (bool) over;
    }

    public Gun()
    {
        this.FIRE_RATE = 0.232f * 3;
        this.MIN_FIRE_SOUND_DURATION = 0.232f * 3;
        this.SIGHTLINE_DISTANCE = 30;
        this.fireSoundDuration = this.MIN_FIRE_SOUND_DURATION;
        this.muzzleCooler = 0.1f;
    }

}